File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_SUPER_PLAYER/SolucionParcialPracticas2/P5-java/src/irrgarten/Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.ArrayList;

/**
 * La clase {@code Player} representa a un jugador en el juego, que posee atributos de
 * inteligencia, fuerza, salud, y una posición en el laberinto. Además, puede equiparse
 * con armas y escudos, atacar y defenderse en el juego.
 * <p>
 * Cada jugador tiene límites en la cantidad de armas y escudos que puede llevar,
 * y también puede resucitar o recibir recompensas durante el juego.
 * </p>
 * 
 * <p> Contiene métodos para mover al jugador, atacar, defender, y recibir heridas.
 * Los jugadores interactúan con otros objetos del juego, como el dado y los elementos
 * de combate, para determinar su rendimiento en el juego.</p>
 * 
 * @author ismael
 */
public class Player extends LabyrinthCharacter{
    /**
     * El número máximo de armas que puede llevar el jugador (valor por defecto: 2).
     */
    private static final int MAX_WEAPONS = 2;

    /**
     * El número máximo de escudos que puede llevar el jugador (valor por defecto: 3).
     */
    private static final int MAX_SHIELDS = 3;

    /**
     * La salud inicial del jugador (valor por defecto: 10).
     */
    private static final int INITIAL_HEALTH = 10;

    /**
     * El número de golpes consecutivos que el jugador puede recibir antes de perder (valor por defecto: 3).
     */
    private static final int HITS2LOSE = 3;

    /**
     * El nombre del jugador.
     */
    private String name;

    /**
     * El número identificador del jugador.
     */
    private char number;

    /**
     * El nivel de inteligencia del jugador.
     */
    private float intelligence;

    /**
     * El nivel de fuerza del jugador.
     */
    private float strength;

    /**
     * El nivel de salud del jugador.
     */
    private int health;

    /**
     * La posición de fila del jugador en el laberinto.
     */
    private int row;

    /**
     * La posición de columna del jugador en el laberinto.
     */
    private int col;

    /**
     * El número de golpes consecutivos recibidos por el jugador.
     */
    private int consecutiveHits = 0;

    /**
     * La lista de armas que lleva el jugador.
     */
    private ArrayList<Weapon> weapons;

    /**
     * La lista de escudos que lleva el jugador.
     */
    private ArrayList<Shield> shields;

    /**
     * El mazo de cartas de armas del jugador.
     */
    private WeaponCardDeck weaponCardDeck;

    /**
     * El mazo de cartas de escudos del jugador.
     */
    private ShieldCardDeck shieldCardDeck;

    /**
     * Constructor para la clase {@code Player}.
     * @param number Identificador del jugador.
     * @param intelligence Nivel de inteligencia del jugador.
     * @param strength Nivel de fuerza del jugador.
     * @param health Nivel de salud del jugador.
     */
    public Player(char number, float intelligence, float strength) {
        super("Player #" + number, intelligence, strength, INITIAL_HEALTH);
        this.number = number;
        this.weapons = new ArrayList<>();
        this.shields = new ArrayList<>();
        this.weaponCardDeck = new WeaponCardDeck();
        this.shieldCardDeck = new ShieldCardDeck();
    }
    
    /**
     * @brief constructor de copia de PlAYER
     * @param other objeto de la clase Player para realizar la copia
     */
    public Player (Player other){
        super(other);
        this.number = other.number;
        this.consecutiveHits = other.consecutiveHits;
        
        this.weapons= new ArrayList<>(other.weapons);
        this.shields= new ArrayList<>(other.shields);

        this.weaponCardDeck = other.weaponCardDeck;
        this.shieldCardDeck = other.shieldCardDeck;

        
        
    }


    /**
     * Resucita al jugador restaurando su salud, hits consecutivos y reiniciando armas y escudos.
     */
    public void resurrect() {    
        setHealth(INITIAL_HEALTH);
        resetHits();
        
        weapons.clear();
        shields.clear();
    }
   

    /**
     * Obtiene el número identificador del jugador.
     * @return El número del jugador.
     */
    public char getNumber() {
        return number;
    }

    /**
     * Calcula el poder de ataque del jugador, basado en su fuerza y el poder de sus armas.
     * @return El poder total de ataque.
     */
    public float attack() {
        return (strength + sumWeapons());
    }

    /**
     * Defiende al jugador de un ataque recibido.
     * @param receivedAttack La potencia del ataque recibido.
     * @return {@code true} si el jugador se defiende exitosamente; {@code false} en caso contrario.
     */
    public boolean defend(float receivedAttack) {
        return manageHit(receivedAttack);
    }

    /**
     * Representación en cadena del jugador.
     * @return Una cadena con los detalles del jugador.
     */
    @Override
    public String toString() {
        String output;
        output = super.toString();
        
        output+=" Consecutive Hits: " + consecutiveHits + "\n";
        output+="\nWeapons: \n";
            for(Weapon w : weapons){
            output+=w.toString();
            output+="\n";
        }
        output+="\nShields: \n";
        for(Shield s : shields){
            output+=s.toString();
            output+="\n";
        }
        return output;
    }

    /**
     * Crea una nueva arma para el jugador.
     * @return Una nueva instancia de {@code Weapon}.
     */
    public Weapon newWeapon() {
        Weapon w = new Weapon(Dice.weaponPower(), Dice.usesLeft());
        return w;
    }

    /**
     * Crea un nuevo escudo para el jugador.
     * @return Una nueva instancia de {@code Shield}.
     */
    public Shield newShield() {
        Shield s = new Shield(Dice.shieldPower(), Dice.usesLeft());
        return s;
    }

    /**
     * Calcula la energía defensiva total de todos los escudos del jugador.
     * @return La energía defensiva acumulada.
     */
    protected float defensiveEnergy() {
        return (intelligence + sumShields());
    }

    /**
     * Reinicia el contador de golpes consecutivos recibidos.
     */
    public void resetHits() {
        consecutiveHits = 0;
    }


    /**
     * Incrementa el contador de golpes consecutivos en 1.
     */
    public void incConsecutiveHits() {
        consecutiveHits++;
    }

    /**
     * Suma la potencia de ataque de todas las armas del jugador.
     * @return El poder total de ataque de todas las armas.
     */
    public float sumWeapons() {
        float sum = 0;
        for (Weapon w : weapons) {
            sum += w.attack();
        }
        return sum;
    }

    /**
     * Suma el poder defensivo de todos los escudos del jugador.
     * @return El poder total defensivo de todos los escudos.
     */
    public float sumShields() {
        float sum = 0;
        for (Shield s : shields) {
            sum += s.protect();
        }
        return sum;
    }

    /**
     * Mueve al jugador en una dirección dada.
     * @param direction La dirección en la que se moverá.
     * @param validMoves Los movimientos válidos disponibles.
     * @return La dirección en la que se movió el jugador.
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves) {
        int size = validMoves.size();
        Directions firstElement;
        boolean contained = false;
        for(Directions d : validMoves){
            if(d==direction) contained=true;
        }
        if(size>0 && !contained){
            firstElement = validMoves.get(0);
        }
        else{
            firstElement=direction;
        }
        return firstElement;
    }

    /**
     * Recibe una recompensa.
     */
    public void receiveReward() {
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        
        for(int i=0; i< wReward; i++){
            Weapon wnew = newWeapon();
            receiveWeapon(wnew);
        }
        for(int i=0; i< sReward; i++){
            Shield snew = newShield();
            receiveShield(snew);
        }
        setHealth(getHealth()+Dice.healthReward());
        
    }
    
    /**
     * Recibe un arma y la agrega al inventario.
     * @param w El arma a recibir.
     */
    public void receiveWeapon(Weapon w) {
        for(int i=0;i<weapons.size();i++){
            Weapon wi = weapons.get(i);
            boolean discard = wi.discard();
            if(discard){
                weapons.remove(i);
            }
        }
        int size = weapons.size();
        if(size<MAX_WEAPONS){
            weapons.add(w);
        }
    }
    
    /**
     * Recibe un escudo y lo agrega al inventario.
     * @param s El escudo a recibir.
     */
    public void receiveShield(Shield s) {
        for(int i=0;i<shields.size();i++){
            Shield si = shields.get(i);
            boolean discard = si.discard();
            if(discard){
                shields.remove(i);
            }
            
        }
        int size = shields.size();
        if(size<MAX_SHIELDS){
            shields.add(s);
        }
    }
    
    /**
     * Gestiona el impacto de un ataque recibido por el jugador.
     * @param receivedAttack La potencia del ataque recibido.
     * @return {@code true} si el impacto fue gestionado exitosamente, {@code false} en caso contrario.
     */
    public boolean manageHit(float receivedAttack) {
        boolean lose=false;
        float defense=defensiveEnergy();
        if(defense<receivedAttack){
            gotWounded();
            incConsecutiveHits();
        }
        else{
            resetHits();
        }
        if(consecutiveHits==HITS2LOSE || dead()){
            resetHits();
            lose=true;
        }
        return lose;
        
    }
    /**
     * Añade una arma a weapons
     * @param w Arma a añadir
     */
    public void addWeapon(Weapon w) {
        weapons.add(w);
    }

    /**
     * Añade un escudo a shields
     * @param s Escudo a añadir
     */
    public void addShield(Shield s) {
        shields.add(s);
    }   
    

}

File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/P5-Ruby/lib/irrgarten/player.rb
#encoding:utf-8

require_relative 'weapon'
require_relative 'shield'
require_relative 'dice'
require_relative 'labyrinth_character'


# Módulo principal del juego Irrgarten, que contiene las clases de jugadores y otros elementos.
module Irrgarten 
    # Representa un jugador en el juego con atributos de salud, posición, fuerza, y defensas.
    class Player < LabyrinthCharacter
        # Número máximo de armas que un jugador puede llevar.
        @@MAX_WEAPONS = 2

        # Número máximo de escudos que un jugador puede llevar.
        @@MAX_SHIELDS = 3

        # Salud inicial con la que un jugador comienza o es resucitado.
        @@INITIAL_HEALTH = 10

        # Cantidad de golpes consecutivos necesarios para perder la partida.
        @@HITS2LOSE = 3

        # Inicializa un nuevo jugador con los atributos básicos de nombre, número, inteligencia y fuerza.
        #
        # @param number [Integer] El número de identificación del jugador.
        def initialize(number, intelligence, strength)
            super("Player#{number}", intelligence, strength, @@INITIAL_HEALTH)
            @number = number.to_s
            @consecutiveHits = 0 # Número de golpes consecutivos recibidos
            @weapons = Array.new # Array de armas
            @shields = Array.new # Array de escudos

        end 

        # Consultor de @number
        # @return [char] número del jugador
        attr_reader :number

        protected
        # Consultor de @weapons
        # @return [Array<Weapon>] array de armas del jugador
        attr_reader :weapons

        # Consultor de @shields
        # @return [Array<Shield>] array de escudos del jugador
        attr_reader :shields

        # Consultor de @consecutive_hits
        # @return [int] número de golpes consecutivos recibidos
        attr_reader :consecutiveHits

        public

        # Constructor de copia
        #
        # @param other [Player] El jugador original del que se creará una copia.
        # @return [void]
        #
        # Copia los atributos básicos del jugador original y los agrega al nuevo jugador.
        def copy(other)
            super
            @number = other.number
            @consecutiveHits = other.consecutiveHits
            @weapons = other.weapons
            @shields = other.shields
        end

        # Resucita al jugador, restaurando su salud inicial y reiniciando sus armas y escudos.
        #
        # @return [void]
        def resurrect 
            @health = @@INITIAL_HEALTH
            resetHits
            @weapons = Array.new
            @shields = Array.new
        end

        # Realiza un ataque utilizando la fuerza del jugador y el poder de sus armas.
        #
        # @return [Integer] La intensidad del ataque.
        def attack
            return @strength + sumWeapons
        end

        # Gestiona la defensa del jugador contra un ataque recibido.
        #
        # @param receivedAttack [Integer] El valor del ataque recibido.
        # @return [Integer] El resultado de la gestión del ataque.
        def defend(receivedAttack)
            return manageHits(receivedAttack)
        end

        # Convierte el estado del jugador en un string para representación.
        #
        # @return [String] La representación del jugador en formato `P[nombre, inteligencia, fuerza, salud, fila, columna]`.
        def to_s
            output = ""
            output += "[PLAYER]" + super.to_s
            output += "\nWeapons:"
            weapons.each do |w|
                output += "\n#{w.to_s}"
            end
            output += "\nShields:"
            shields.each do |s|
                output += "\n#{s.to_s}"
            end
            output += "\nConsecutive hits: #{@consecutiveHits}\n"
        end


        
        # Mueve al jugador en una dirección específica si es válida.
        #
        # @param direction [String] La dirección en la que se desea mover.
        # @param validMoves [Array<String>] Las direcciones válidas en las que el jugador puede moverse.
        # @return [String] La dirección en la que el jugador se moverá.
        def move(direction, validMoves) 
            size = validMoves.size
            firstElement = direction
            contained = validMoves.include?(direction)
            
            if size > 0 && !contained
                firstElement = validMoves[0]
            end
                
            firstElement
        end
        
        # Recibe recompensas para el jugador, que pueden incluir armas, escudos y salud extra.
        # El número de armas y escudos recibidos se determina lanzando dados.
        # La salud extra también se determina lanzando un dado.
        #
        # @return [void]
        def receiveReward
            wReward = Dice.weaponReward
            sReward = Dice.shieldReward

            for i in 0...wReward
                wnew = newWeapon
                receiveWeapon(wnew)
            end

            for i in 0...sReward
                snew = newShield
                receiveShield(snew)
            end

            extraHealth = Dice.healthReward
            @health += extraHealth
        end
        
        private 

        # Recibe una nueva arma para el jugador, descartando las armas usadas si es necesario.
        #
        # @param w [Weapon] La nueva arma a recibir.
        # @return [void]
        def receiveWeapon(w)
            @weapons.delete_if do |wi| # eliminamos las armas que debemos de eliminar
                wi.discard
            end

            if(@weapons.length < @@MAX_WEAPONS) # si caben se añaden
                @weapons.push(w)
            end
        end

        # Recibe un nuevo escudo para el jugador, descartando los escudos usados si es necesario.
        #
        # @param s [Shield] El nuevo escudo a recibir.
        # @return [void]    
        def receiveShield(s)
            @shields.delete_if do |si| # eliminamos los escudos que debemos de eliminar
                si.discard
            end

            if(@shields.length < @@MAX_SHIELDS) # si caben se añaden
                @shields.push(s)
            end
        end

        # Crea una nueva arma con poder y usos restantes aleatorios.
        #
        # @return [Weapon] La nueva arma generada.
        def newWeapon
            w = Weapon.new(Dice.weaponPower, Dice.usesLeft)
            return w 
        end

        # Crea un nuevo escudo con poder defensivo aleatorio.
        #
        # @return [Shield] El nuevo escudo generado.
        def newShield
            s = Shield.new(Dice.shieldPower, Dice.usesLeft)
            return s
        end
        

        # Gestiona los golpes recibidos por el jugador y determina si pierde la partida.
        #
        # @param receivedAttack [Integer] El valor del ataque recibido.
        # @return [Boolean] `true` si el jugador pierde la partida, `false` en caso contrario.
        def manageHits(receivedAttack)
            lose = true
            defense = defensiveEnergy
            if defense < receivedAttack
                gotWounded
                incConsecutiveHits
            else
                resetHits
            end
            if @consecutiveHits == @@HITS2LOSE || dead 
                resetHits
                lose = true
            else
                lose = false
            end
            return lose
        end

        # Reinicia el contador de golpes consecutivos del jugador.
        #
        # @return [void]
        def resetHits
            @consecutiveHits = 0
        end

        # Incrementa el contador de golpes consecutivos del jugador en 1.
        #
        # @return [void]
        def incConsecutiveHits
            @consecutiveHits += 1
        end

        protected 

        # Suma el poder de ataque de todas las armas del jugador.
        #
        # @return [Integer] La suma del poder de todas las armas.
        def sumWeapons 
            sum = 0
            @weapons.each do |w|
                sum += w.attack
            end
            sum
        end

        # Suma la energía defensiva de todos los escudos del jugador.
        #
        # @return [Integer] La suma de la energía defensiva de todos los escudos.
        def sumShields 
            sum = 0
            @shields.each do |s|
                sum += s.protect
            end
            sum
        end

        # Calcula la energía defensiva total del jugador sumando el poder de todos sus escudos.
        #
        # @return [Integer] La energía defensiva total.
        def defensiveEnergy
            return @intelligence + sumShields
        end

        





    end
end